### ISSUE TITLE
AdaptiveScheduler collapses per-(B,H) entropy to one scalar, defeating its purpose

### ISSUE THREAD

**Issue by sachncs** [sachncs/avqa#15]
## Problem

`src/avqa/scheduler.py:113-124` `AdaptiveScheduler.budget_for` averages the importance tensor across all batch and head dims before computing entropy:

```python
def budget_for(self, importance: torch.Tensor) -> int:
    flat = importance.mean(dim=tuple(range(importance.ndim - 1)))  # [M_0]
    flat = flat / flat.sum().clamp_min(1e-12)
    entropy = -(flat * flat.clamp_min(1e-12).log()).sum().item()
    max_entropy = float(torch.log(torch.tensor(flat.numel())).item())
    norm_entropy = entropy / max_entropy if max_entropy > 0 else 1.0
    if norm_entropy < self.entropy_threshold:
        return self.max_budget
    return self.min_budget
```

This produces **one scalar** budget for the whole batch × heads, ignoring per-(B, H) variation.

## Evidence

- `src/avqa/scheduler.py:113-118` (quoted above).
- `src/avqa/pipeline.py:340` calls `budget = state.scheduler.budget_for(importance)` and then `min(budget, min_valid)` at line 343, so a single scalar budget is applied to every (B, H) pair.
- `CHANGELOG.md` "Unreleased / Fixed" line 60-63 explicitly acknowledges this: *"**`AdaptiveScheduler` collapsed per-(B, H) entropy to a single scalar** (routing-decoupled, kept as noted; deferred for next minor)."*
- The docstring (`scheduler.py:79-91`) promises "Entropy-driven adaptive budget" with no caveat about the collapse.

## Root Cause

The author chose `mean(dim=tuple(range(importance.ndim - 1)))` instead of an elementwise entropy that could return a per-(B, H) budget.

## Impact

- The "adaptive" scheduler is functionally identical to a fixed-budget scheduler for any practical workload: every batch uses the same budget. The name and behavior diverge.
- Users that set `RefinementConfig(adaptive_budget=True)` get no routing benefit; the only practical difference from `adaptive_budget=False` is the choice of `min_budget = max(1, budget // 2)` (line 41).
- The scheduler passes its unit tests (`tests/unit/test_scheduler.py::TestAdaptiveScheduler`) only because those tests assert `budget_for(...)` returns a scalar.

## Recommendation

Change the return type of `Scheduler.budget_for` to `int | torch.Tensor` (Tensor of shape `[B, H]`), or to an explicit per-(B, H) tuple. Update `refinement.py` and `pipeline.py` to consume the per-(B, H) budget and clamp each (B, H) against its own `num_valid_per_bh`.

## Acceptance Criteria

- [ ] `AdaptiveScheduler.budget_for` returns a budget per (B, H) position.
- [ ] `refine_and_output` clamps `budget` per (B, H), not as a scalar.
- [ ] Add a test that two importance distributions with different per-(B, H) entropies yield different per-(B, H) budgets.
- [ ] `DefaultScheduler` continues to return a scalar (or a broadcastable tensor).

## Verification

`pytest tests/unit/test_scheduler.py -v` passes, plus a new property test:
```python
def test_adaptive_budget_responds_per_bh():
    s = AdaptiveScheduler(min_budget=2, max_budget=8)
    importance_focused = torch.zeros(2, 4, 16); importance_focused[..., 0] = 100
    importance_uniform = torch.ones(2, 4, 16)
    # New contract: per-(B, H) budget tensor.
    out_focused = s.budget_for(importance_focused)
    out_uniform = s.budget_for(importance_uniform)
    assert (out_focused >= out_uniform).all()  # focused -> more budget
```

## Risk

Low (the scheduler is wired in only by `AVQAttention` and one orchestrator call site). The change is internal; tests will pin the new contract.

**Comment by sachncs** [comment]
Resolved in PR #37.
